src/cls/journal/cls_journal.cc

```diff
@@ -224,6 +224,86 @@ int expire_tags(cls_method_context_t hctx, const std::string *skip_client_id) {
   return 0;
 }
 
+int get_client_list_range(cls_method_context_t hctx,
+                          std::set<cls::journal::Client> *clients,
+                          std::string start_after, uint64_t max_return) {
+  std::string last_read;
+  if (!start_after.empty()) {
+    last_read = key_from_client_id(start_after);
+  }
+
+  std::map<std::string, bufferlist> vals;
+  int r = cls_cxx_map_get_vals(hctx, last_read, HEADER_KEY_CLIENT_PREFIX,
+                               max_return, &vals);
+  if (r < 0) {
+    CLS_ERR("failed to retrieve omap values: %s", cpp_strerror(r).c_str());
+    return r;
+  }
+
+  for (std::map<std::string, bufferlist>::iterator it = vals.begin();
+       it != vals.end(); ++it) {
+    try {
+      bufferlist::iterator iter = it->second.begin();
+
+      cls::journal::Client client;
+      ::decode(client, iter);
+      clients->insert(client);
+    } catch (const buffer::error &err) {
+      CLS_ERR("could not decode client '%s': %s", it->first.c_str(),
+              err.what());
+      return -EIO;
+    }
+  }
+
+  return 0;
+}
+
+int find_min_commit_position(cls_method_context_t hctx,
+                             cls::journal::ObjectSetPosition *minset) {
+  int r;
+  bool valid;
+  uint64_t max_return;
+  uint64_t tag_tid, entry_tid;
+  std::string start_after;
+  std::set<cls::journal::Client> batch;
+  cls::journal::ObjectPosition first;
+  cls::journal::ObjectSetPosition object_set_position;
+
+  valid = false;
+  tag_tid = entry_tid = 0;
+
+  start_after = "";
+  max_return = cls::journal::JOURNAL_MAX_RETURN;
+
+  do {
+    batch.clear();
+    r = get_client_list_range(hctx, &batch, start_after, max_return);
+    if ((r == 0) && !batch.empty()) {
+      start_after = batch.rbegin()->id;
+
+      // update the (minimum) commit position from this batch of clients
+      for(std::set<cls::journal::Client>::iterator it = batch.begin();
+          it != batch.end(); ++it) {
+        object_set_position = (*it).commit_position;
+        if (!object_set_position.object_positions.empty()) {
+          first = object_set_position.object_positions.front();
+
+          // least tag_tid (or least entry_tid for matching tag_tid)
+          if (!valid || (tag_tid > first.tag_tid)
+              || ((tag_tid == first.tag_tid) && (entry_tid > first.entry_tid))) {
+            tag_tid = first.tag_tid;
+            entry_tid = first.entry_tid;
+            *minset = cls::journal::ObjectSetPosition(object_set_position);
+            valid = true;
+          }
+        }
+      }
+    }
+  } while((r == 0) && !(batch.size() < max_return));
```

Comment:
I was just commenting on the negation of a less-than operator -- it looks a bit confusing.  We should know we can break out of the loop if we had an error (won't actually be necessary if we break out above) or if we retrieved less than the requested number of clients.